### PR TITLE
New schema

### PR DESCRIPTION
**Pull request by wpears**
After #133:

Loads both combined address string and individual address components.

### CODE REVIEW

File: data.json [hunk 1]
```diff
@@ -52,13 +61,17 @@
...
-      "Address": {
-        "type": "static",
-        "value": "AddressLabel"
+      "Number": {
+        "type": "multi",
+        "value": ["PrefixAddressNumber", "AddressNumber", "SuffixAddressNumber"]
+      },
+      "Street": {
+        "type": "multi",
+        "value": ["PreModifier", "PreDirectional", "PreType", "StreetName", "PostType", "PostDirectional", "PostModifier"]
```
Comment (hkeeler) [inline data.json]: Wow.  This schema matches up exactly with the usaddress parser.  They must be using the same standard (I forget what this is).

I'm doing my mapping in [`rules.yaml`](https://github.com/cfpb/grasshopper-parser/pull/33/files#diff-b18b554cc815926f6d91670d1aaca05bR125), and it's _almost_ the same order, just the first two elements are switched.  I'm currently using the order they occur in the address string, not the order of the list, but that _could_ change.  Do you know for sure that PreModifier should come before PreDirectional?  If so, I can change mine.

```yaml
    - id: &street_name_full street_name_full
      description: 'a modifier adter a street name, e.g. "Ext"'
      parts:
        - *street_name_pre_directional
        - *street_name_pre_modifier
        - *street_name_pre_type
        - *street_name
        - *street_name_post_type
        - *street_name_post_directional
        - *street_name_post_modifier
```

Huh, also looks like my `description` above is both incorrect and has a misspelling.  Guess I should fix that.

File: lib/resolveFields.js [hunk 2]
```diff
@@ -1,16 +1,36 @@
...
     var field = fields[keys[i]];
     if(field.type === 'dynamic'){
       val = new Function('props', field.value)(props); //eslint-disable-line
+    }else if(field.type === 'multi'){
+      val = field.value.map(mapProps).filter(truthy).join(' ');
     }else{
       val = props[field.value];
     }
+
+    val = (val + '').trim();
```
Comment (hkeeler) [inline lib/resolveFields.js]: Why is this necessary?  Is this some `null`-to-string or magic for the `val = Function`?

File: lib/resolveFields.js [hunk 3]
```diff
@@ -1,16 +1,36 @@
...
+
     vals[keys[i]] = val;
   }
 
   return vals;
 }
+
+
+function truthy(v){
```
Comment (hkeeler) [inline lib/resolveFields.js]: I'm confused.  I see this function is passed into the `.filter()` above, but...?  What does this do?

**Comment by hkeeler** [comment]
All reviewed up.  Just a few questions sprinkled in there...mostly JS head-scratchers I think.

Also, looks like Travis stalled out.  Not sure if that's in any way related to this PR:
>`No output has been received in the last 10 minutes, this potentially indicates a stalled build or something wrong with the build itself.`

Once #133 is merged and Travis is happy, I'll give it another once over and merge.

File: data.json (hunk 1 above)
Comment (wpears) [inline data.json]: In the json object that gets produced by ogr2ogr on their raw data, the pre modifier comes first.

This is reflected in the following row address:
````
{
  "type": "Feature",
  "properties": {
    "NYSAddressPointID": "WYOM000548",
    "CountyID": "6180",
    "NYSStreetID": 437252580,
    "PrefixAddressNumber": null,
    "AddressNumber": 6860,
    "SuffixAddressNumber": null,
    "PreModifier": "Old",
    "PreDirectional": "E",
    "PreType": null,
    "SeparatorElement": null,
    "StreetName": "Arcade",
    "PostType": "Rd",
    "PostDirectional": null,
    "PostModifier": null,
    "SubAddress": null,
    "LandmarkName": null,
    "Building": null,
    "Floor": null,
    "Unit": null,
    "Room": null,
    "Seat": null,
    "Location": null,
    "ZipName": "Arcade",
    "State": "NY",
    "ZipCode": "14009",
    "PointType": "1",
    "AddressSource": "County AP",
    "SourceOfData": null,
    "DateUpdated": "2014/01/29 00:00:00",
    "PrimaryPoint": "Y",
    "CityTownName": "Arcade",
    "PlaceName": null,
    "PlaceType": null,
    "CompleteStreetName": "Old East Arcade Road",
    "AP_Flag": null,
    "Status": "Active",
    "AddressLabel": "6860 Old E Arcade Rd \r\n",
    "CR_UserName": null,
    "CR_DateTime": null,
    "AT_UserName": null,
    "AT_DateTime": null,
    "SP_UserName": null,
    "SP_DateTime": null,
    "CountyName": "Wyoming",
    "CLDXF_PreDirectional": "East",
    "CLDXF_PostType": "Road",
    "CLDXF_PostDirectional": null,
    "Latitude": 42.57014195419415,
    "Longitude": -78.34871794629433
  },
  "geometry": {
    "type": "Point",
    "coordinates": [
      225143.04760000017,
      4718518.536
    ]
  }
}
````

Where premodifier is ````Old```` and predirectional is ````E```` leading to ````6860 Old E Arcade Rd```` which intuitively sounds right to me.

Thoughts?

File: lib/resolveFields.js (hunk 2 above)
Comment (wpears) [inline lib/resolveFields.js]: Sorry this is toString hackery. Adding empty string to most things is equivalent to calling ````whatever.toString()````. This happens to be both shorter and faster... but also less clear.

File: lib/resolveFields.js (hunk 3 above)
Comment (wpears) [inline lib/resolveFields.js]: The filter method of an array will call the provided function on every entry and construct a new array containing the each entry for which the function returns a "truthy" value (javascript will cast the return value to a boolean and then check true/false). By passing a function that just returns the provided value, we'll be left with only truthy arguments in a new array.

In practice, this is used to munge data that is either null or empty string.

File: data.json (hunk 1 above)
Comment (hkeeler) [inline data.json]: Here's where it gets really interesting.  The usaddress parser has both!  The following are in usaddress's [`labeled.xml`](https://github.com/datamade/usaddress/blob/master/training/labeled.xml) training file:

**[Exhibit A: `PreModifier` -> `PreDirectional`](https://github.com/datamade/usaddress/blob/master/training/labeled.xml#L972)**

    5442 OLD WEST STATE ROUTE 21,IMPERIAL, MO 63052

parses to:

```xml
<AddressString><AddressNumber>5442</AddressNumber> <StreetNamePreModifier>OLD</StreetNamePreModifier> <StreetNamePreDirectional>WEST</StreetNamePreDirectional> <StreetNamePreType>STATE</StreetNamePreType> <StreetNamePreType>ROUTE</StreetNamePreType> <StreetName>21,</StreetName> <PlaceName>IMPERIAL,</PlaceName> <StateName>MO</StateName> <ZipCode>63052</ZipCode></AddressString>
```

**[Exhibit B: `PreDirectional` -> `PreModifier`](https://github.com/datamade/usaddress/blob/master/training/labeled.xml#L859)**

    135 W OLD US HIGHWAY 90

parses to:

```xml
<AddressString><AddressNumber>135</AddressNumber> <StreetNamePreDirectional>W</StreetNamePreDirectional> <StreetNamePreModifier>OLD</StreetNamePreModifier> <StreetNamePreType>US</StreetNamePreType> <StreetNamePreType>HIGHWAY</StreetNamePreType> <StreetName>90</StreetName></AddressString>
```

So, for now, I think its safe to just keep things as they are, and the parser will always just return the values as they occur in the address string.

**Comment by hkeeler** [comment]
Looks good.  Merging...

File: lib/resolveFields.js (hunk 3 above)
Comment (hkeeler) [inline lib/resolveFields.js]: Wat!
